### PR TITLE
RAB-967: Write service API

### PR DESCRIPTION
**Pull request by MadeWilson**
add Service Api to list Users

**Definition Of Done (for Core Developer only)**

- [x] Tests

### CODE REVIEW

File: src/Akeneo/UserManagement/back/Application/Handler/ListUsersHandler.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+final class ListUsersHandler implements ListUsersHandlerInterface
+{
+    public function __construct(
+        private FindUsers $findUsers
+    ) {
+    }
+
+    public function fromQuery(ListUsersQuery $query): array
+    {
+        return ($this->findUsers)(
```
Comment (StevenVAIDIE) [inline src/Akeneo/UserManagement/back/Application/Handler/ListUsersHandler.php:29]: We should not return a Domain object in a ServiceAPI

File: src/Akeneo/UserManagement/back/Infrastructure/Storage/SqlFindUsers.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        $query = $this->buildQuery($search, $searchAfterId, $limit);
+
+        $results = $this->connection->executeQuery($query)->fetchAllAssociative();
+
+        return array_map(
+            static function ($data) {
+                return ServiceAPiUser::createFromDatabase($data);
+            },
+            $results
+        );
```
Comment (StevenVAIDIE) [inline src/Akeneo/UserManagement/back/Infrastructure/Storage/SqlFindUsers.php:48]:
```suggestion
        return array_map(static fn ($data) => ServiceAPiUser::createFromDatabase($data), $results);
```

File: src/Akeneo/UserManagement/back/Infrastructure/Storage/SqlFindUsers.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    private function buildQuery(?string $search, ?int $searchAfterId, ?int $limit): string
+    {
+        $searchSql = '';
+        if (null !== $search) {
+            $searchSql = <<<SQL
+                AND (ou.username LIKE '%${search}%' OR ou.email LIKE '%${search}%' OR ou.first_name LIKE '%${search}%' OR ou.last_name LIKE '%${search}%')
```
Comment (StevenVAIDIE) [inline src/Akeneo/UserManagement/back/Infrastructure/Storage/SqlFindUsers.php:56]: Injection SQL possible here

File: src/Akeneo/UserManagement/back/Infrastructure/Storage/SqlFindUsers.php [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            SQL;
+        }
+
+        $searchAfterIdSql = '';
+        if (null !== $searchAfterId) {
+            $searchAfterIdSql = <<<SQL
+                AND ou.id > ${searchAfterId}
```
Comment (StevenVAIDIE) [inline src/Akeneo/UserManagement/back/Infrastructure/Storage/SqlFindUsers.php:63]: Same

File: src/Akeneo/UserManagement/back/ServiceApi/User/ListUsersQuery.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Akeneo\UserManagement\ServiceApi\User;
+
+use Akeneo\UserManagement\Domain\Storage\FindUsers;
+
+final class ListUsersQuery
```
Comment (StevenVAIDIE) [inline src/Akeneo/UserManagement/back/ServiceApi/User/ListUsersQuery.php:18]: Like ListUserGroupInterface we should use a generic query related to the User 
```suggestion
final class UsersQuery
```

File: src/Akeneo/UserManagement/back/tests/Integration/Infrastructure/Storage/SqlFindUsersIntegration.php [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        $this->createUserWithGroupsAndRoles(3, 'test3', ['Redactor'], ['ROLE_USER']);
+        $this->createUserWithGroupsAndRoles(4, 'julia', ['Redactor'], ['ROLE_USER']);
+    }
+
+    public function testItListsTheUsers(): void
+    {
+        $users = ($this->get(SqlFindUsers::class))();
+
+        Assert::assertCount(4, $users);
+        Assert::containsOnlyInstancesOf(User::class);
```
Comment (StevenVAIDIE) [inline src/Akeneo/UserManagement/back/tests/Integration/Infrastructure/Storage/SqlFindUsersIntegration.php:42]: Should we validate the roles, the groups, email an other information on the User ?

File: src/Akeneo/UserManagement/back/tests/Integration/ServiceApi/User/ListUsersQueryIntegration.php [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        $this->createUserWithGroupsAndRoles(3, 'test3', ['Redactor'], ['ROLE_USER']);
+        $this->createUserWithGroupsAndRoles(4, 'julia', ['Redactor'], ['ROLE_USER']);
+    }
+
+    public function testItListsTheUsers(): void
+    {
+        $users = $this->getHandler()->fromQuery(new ListUsersQuery());
+
+        Assert::assertCount(4, $users);
+        Assert::containsOnlyInstancesOf(User::class);
```
Comment (StevenVAIDIE) [inline src/Akeneo/UserManagement/back/tests/Integration/ServiceApi/User/ListUsersQueryIntegration.php:43]: It should not be the Domain object

File: src/Akeneo/UserManagement/back/tests/Integration/Infrastructure/Storage/SqlFindUsersIntegration.php (hunk 6 above)
Comment (MadeWilson) [reply src/Akeneo/UserManagement/back/tests/Integration/Infrastructure/Storage/SqlFindUsersIntegration.php:42]: we didn't search for a user with a specific role or group the api can't do that as of now so not sure for this two, email could deserve a test has it is searchable.

File: src/Akeneo/UserManagement/back/ServiceApi/User/UsersQuery.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+namespace Akeneo\UserManagement\ServiceApi\User;
+
+use Akeneo\UserManagement\Domain\Storage\FindUsers;
+
+final class UsersQuery
+{
+    public function __construct(
+        private ?string $searchName = null,
```
Comment (grena) [inline src/Akeneo/UserManagement/back/ServiceApi/User/UsersQuery.php:21]:
```suggestion
        private ?string $search = null,
```
As we also search in email, username, etc... `$search` is more appropriate I think.

File: src/Akeneo/UserManagement/back/tests/Integration/ServiceApi/User/ListUsersHandlerIntegration.php [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        Assert::containsOnlyInstancesOf(User::class);
+        Assert::assertEquals('julia', $users[0]->getUsername());
+    }
+
+    public function testItListsTheUsersWithPagination(): void
+    {
+        $users = $this->getHandler()->fromQuery(new UsersQuery(
+            null,
+            null,
+            2
```
Comment (grena) [inline src/Akeneo/UserManagement/back/tests/Integration/ServiceApi/User/ListUsersHandlerIntegration.php:60]: You can use named parameters in tests, it's more readable to see what we're searching

**Review by grena — APPROVED** [review]

File: src/Akeneo/UserManagement/back/Infrastructure/Storage/SqlFindUsers.php [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        $results = $this->connection->executeQuery(
+            $query,
+            [
+                'search' => sprintf('%%%s%%', $search),
+                'searchAfterId' => $searchAfterId,
+            ],
+        )->fetchAllAssociative();
+
+        return array_map(
+            static fn ($data) => ServiceAPiUser::createFromDatabase($data),
```
Comment (axelhms) [inline src/Akeneo/UserManagement/back/Infrastructure/Storage/SqlFindUsers.php:50]: Small typo
```suggestion
            static fn ($data) => ServiceApiUser::createFromDatabase($data),
```

File: src/Akeneo/UserManagement/back/ServiceApi/User/User.php [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        return $this->username;
+    }
+
+    public function getUserType(): string
+    {
+        return $this->userType;
+    }
+
+    public function getFirstname(): string
```
Comment (axelhms) [inline src/Akeneo/UserManagement/back/ServiceApi/User/User.php:51]: Can also return null, same for the others underneath
```suggestion
    public function getFirstname(): ?string
```

**Review by axelhms — APPROVED** [review]

**Pull request merged**